src/richie/apps/search/utils/filter_definitions.py

```diff
@@ -0,0 +1,259 @@
+"""
+Define our default FilterDefinition classes.
+They encapsulate common behavior around managing filters in ElasticSearch queries and
+API requests (to validate input) and responses (to produce easy-to-consume filters).
+"""
+from functools import reduce
+
+from django.conf import settings
+from django.utils.module_loading import import_string
+
+from .i18n import get_best_field_language
+
+
+class FilterDefinitionBase:
+    """
+    Base filter definition shape: just an init method to set common attributes.
+    """
+
+    def __init__(self, name, human_name, **kwargs):
+        """
+        Assign the common attributes and set some defaults for optional parameters.
+        """
+        self.name = name
+        self.human_name = human_name
+        # Set default values for optional parameters
+        self.is_drilldown = kwargs.get("is_drilldown") or False
+
+
+class FilterDefinitionCustom(FilterDefinitionBase):
+    """
+    Filter definition for a custom filter with hardcoded choices provided along with their
+    ES filtering fragment.
+    """
+
+    def __init__(self, name, human_name, choices, **kwargs):
+        """
+        Extend the base filter definition with tools to easily generate query/aggregation
+        fragments et faceted filter definition.
+        """
+        super().__init__(name, human_name, **kwargs)
+        # The key/fragment map is static and helpful to get query & aggregations fragments
+        self._key_fragment_map = {choice[0]: choice[2] for choice in choices}
+        # Extend the base filter definition with a list of values that includes names & human
+        # names and is only missing the counts.
+        # We do this because:
+        #   - the number of choices is bounded as it is a static argument (as opposed to eg.)
+        #     a row in the database;
+        #   - it makes it easier to access the values by key with every search request if we
+        #     use an dict (while keeping a simpler interface for richie lib users).
+        self.values = {
+            value_name: {"human_name": value_human_name, "key": value_name}
+            for value_name, value_human_name, _ in choices
```

Comment:
Because `values` are fundamentally a list, including when we return them from the API; we'd have to rebuild the list below to return it to the client.

IMO it's better to not have separate formats for the exact same data in different places.

Reply:
ok then.